src/ess/reduce/streaming.py

```diff
@@ -0,0 +1,213 @@
+        self._value: T | None = None
+
+    @property
+    def value(self) -> T:
+        return self._value.copy()
+
+    def _do_push(self, value: T) -> None:
+        if self._value is None:
+            self._value = value.copy()
+        else:
+            self._value += value
+
+
+class RollingAccumulator(Accumulator[T]):
+    """
+    Accumulator that adds pushed values to a rolling window.
+
+    Does not support event data.
+    """
+
+    def __init__(self, window: int = 10, **kwargs: Any) -> None:
+        """
+        Parameters
+        ----------
+        window:
+            Size of the rolling window.
+        """
+        super().__init__(**kwargs)
+        self._window = window
+        self._values: list[T] = []
+
+    @property
+    def value(self) -> T:
+        # Naive and potentially slow implementation if values and/or window are large!
+        return sc.reduce(self._values).sum()
+
+    def _do_push(self, value: T) -> None:
+        self._values.append(value)
+        if len(self._values) > self._window:
+            self._values.pop(0)
+
+
+class StreamProcessor:
+    """
+    Wrap a base workflow for streaming processing of chunks.
+
+    Note that this class can not determine if the workflow is valid for streamed
+    processing based on the input keys. In particular, it is the responsibility of the
+    user to ensure that the workflow is "linear" with respect to the dynamic keys up to
+    the accumulation keys.
+    """
+
+    def __init__(
+        self,
+        base_workflow: sciline.Pipeline,
+        dynamic_keys: tuple[sciline.typing.Key, ...],
+        accumulation_keys: tuple[sciline.typing.Key, ...],
+        target_keys: tuple[sciline.typing.Key, ...],
+        accumulator: type[Accumulator] = EternalAccumulator,
+    ) -> None:
+        """
+        Create a stream processor.
+
+        Parameters
+        ----------
+        base_workflow:
+            Workflow to be used for processing chunks.
+        dynamic_keys:
+            Keys that are expected to be updated with each chunk.
+        accumulation_keys:
+            Keys for which to accumulate values.
+        target_keys:
+            Keys to be computed and returned.
+        accumulator:
+            Accumulator class to use for accumulation.
+        """
+        workflow = sciline.Pipeline()
+        for key in target_keys:
+            workflow[key] = base_workflow[key]
+        for key in dynamic_keys:
+            workflow[key] = None  # hack to prune branches
+
+        # Find and pre-compute static nodes as far down the graph as possible
+        # See also https://github.com/scipp/sciline/issues/148.
+        nodes = _find_descendants(workflow, dynamic_keys)
+        parents = _find_parents(workflow, nodes) - nodes
+        for key, value in base_workflow.compute(parents).items():
+            workflow[key] = value
+
+        self._process_chunk_workflow = workflow.copy()
+        self._finalize_workflow = workflow.copy()
```

Comment:
Is there a simple way to split the graph at the accumulated nodes? It took me a moment to understand why there are two copies of the same workflow. Plus, splitting would reduce the risk of misusing the workflows.

Reply:
This does look very simple: https://stackoverflow.com/questions/65853641/networkx-how-to-split-nodes-into-two-effectively-disconnecting-edges

I have previously considered extending the slicing syntax to allow for specifying start and end using `graph[A:B]`, but that is not implemented yet. Also, I would not work if multiple starts/ends are required (unless we want to support a list syntax, `graph[A:[B1,B2]]`).